addons/website_sale_picking/models/delivery_carrier.py

```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields, _, api
+from odoo.exceptions import ValidationError
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = 'delivery.carrier'
+
+    # Onsite delivery means the client comes to a physical store to get the products himself.
+    delivery_type = fields.Selection(selection_add=[
+        ('onsite', 'Pickup in store')
+    ], ondelete={'onsite': 'set default'})
+
+    # If set, the sales order shipping address will take this warehouse's address.
+    warehouse_id = fields.Many2one('stock.warehouse', 'Warehouse')
+
+    @api.constrains('warehouse_id', 'company_id')
+    def _check_warehouse_company(self):
+        for record in self:
+            if record.warehouse_id and record.warehouse_id.company_id:
+                if record.company_id and record.company_id != record.warehouse_id.company_id:
+                    raise ValidationError(_("The picking site and warehouse must share the same company"))
```

Comment:
Accessing a field on an empty recordset don't crash actually. You don't have to make all those checks *à la* Javascript. On the other hand, we try to avoid being abstract when naming variable (Better to say ``for carrier in self:`` for example)
```suggestion
        for carrier in self:
            if carrier.warehouse_id.company_id and carrier.company_id != carrier.warehouse_id.company_id:
                    raise ValidationError(_("The picking site and warehouse must share the same company"))
```